### PR TITLE
Fix/replace refs for relationships

### PR DESCRIPTION
**Pull request by barnabasJ**
# Contributor checklist

- [ ] Bug fixes include regression tests
- [ ] Chores
- [ ] Documentation changes
- [ ] Features include unit/acceptance tests
- [ ] Refactoring
- [ ] Update dependencies

### CODE REVIEW

File: lib/ash/policy/authorizer/authorizer.ex [hunk 1]
```diff
@@ -712,6 +712,25 @@ defmodule Ash.Policy.Authorizer do
...
+         } = ref,
+         %{stack: [{parent, _path, _action} | _]} = acc
+       )
+       when struct in [Ash.Resource.Attribute, Ash.Resource.Aggregate, Ash.Resource.Calculation] do
+    action =
+      Ash.Resource.Info.relationship(parent, relationship_path).read_action ||
+        Ash.Resource.Info.primary_action!(resource, :read)
+
+    {expr, acc} =
+      expression_for_ref(resource, name, action, ref, acc)
```
Comment (barnabasJ) [inline lib/ash/policy/authorizer/authorizer.ex:729]: We might also need to update the stack here. But doing it caused an error for me. 

I hit this case when coming from the `%{__function__?: true, ... }` above. Returning the original stack there might be the correct way to do things. What do you think?

Comment (zachdaniel) [reply lib/ash/policy/authorizer/authorizer.ex:729]: 🤔 I'm pretty sure the way you have it here is the right way :)

**Comment by zachdaniel** [comment]
This looks good, *but* I don't understand exactly how we got to the scenario, a failing test would be great. If you have time. If not, we can merge.

**Review by zachdaniel — DISMISSED** [review]

**Comment by zachdaniel** [comment]
I think it should be fine because `expression_for_ref` doesn't use the `stack` IIRC.

**Comment by barnabasJ** [comment]
I tried adding a test. I did it in AshPostgres because I needed a filter function e.g `ilke` to be available. But I'm still having a hard time triggering the problem. Right now it does not call do_replace_refs because the `input?` flag on the Ref is not true. How do I need to set the filter to make it so?

**Comment by barnabasJ** [comment]
> I tried adding a test. I did it in AshPostgres because I needed a filter function e.g `ilke` to be available. But I'm still having a hard time triggering the problem. Right now it does not call do_replace_refs because the `input?` flag on the Ref is not true. How do I need to set the filter to make it so?

I was able to figure it out by looking at the ash_graphql resolver, still having issues replicating the error though

**Comment by barnabasJ** [comment]
I added this test in ash_postgres https://github.com/ash-project/ash_postgres/pull/206

The behaviour is a bit strange over all, but this way, you should be able to get a better picture of it.

**Comment by barnabasJ** [comment]
> I added this test in ash_postgres [ash-project/ash_postgres#206](https://github.com/ash-project/ash_postgres/pull/206)
> 
> The behaviour is a bit strange over all, but this way, you should be able to get a better picture of it.

I added another case that is two relationships deep. This still fails even with my current change.

File: lib/ash/policy/authorizer/authorizer.ex [hunk 2]
```diff
@@ -712,6 +712,25 @@ defmodule Ash.Policy.Authorizer do
     end
   end
 
+  defp do_replace_ref(
+         %{
+           attribute: %struct{name: name},
+           relationship_path: [relationship_path],
```
Comment (zachdaniel) [inline lib/ash/policy/authorizer/authorizer.ex:718]:
```suggestion
           relationship_path: relationship_path,
```

**Comment by zachdaniel** [comment]
🚀 Thank you for your contribution! 🚀

This looked like some hard work to narrow this done. Fixed one small issue in your PR but otherwise this is good to go!

**Pull request merged**